### PR TITLE
Rewrite Map to Filter inlining in `PredicatePushdown`

### PR DESCRIPTION
**Pull request by ggevay**
This addresses https://github.com/MaterializeInc/materialize/issues/22040 by changing the condition of when to inline a Map expression into a Filter expression in `PredicatePushdown` to be less restrictive, as explained in the issue description, and makes the inlining more general to support transitive inlining through several Map expressions.

The first commit just adds a `reset-server` in an slt as discussed [here](https://materializeinc.slack.com/archives/C01LKF361MZ/p1696259071624369?thread_ts=1696258870.495249&cid=C01LKF361MZ).

### Motivation

  * This PR fixes a recognized bug: https://github.com/MaterializeInc/materialize/issues/22040

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
  - [ ] New Nightly run: https://buildkite.com/materialize/nightlies/builds/4611
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/transform/src/predicate_pushdown.rs [hunk 1]
```diff
@@ -897,50 +898,109 @@ impl PredicatePushdown {
...
         Ok((retained, pushdown))
     }
 
+    /// This fn should be called with a Filter `expr` that is after a Map. `input_arity` is the
+    /// arity of the input of the Map. This fn eliminates such column refs in `expr` that refer not
+    /// to a column in the input of the Map, but to a column that is created by the Map. It does
+    /// this by transitively inlining Map expressions until no such expression remains that points
+    /// to a Map expression. The return value is the cleaned up expression. The fn bails out with a
+    /// None if the resulting expression would be made too big by the inlinings.
+    fn inline_if_not_too_big(
```
Comment (aalexandrov) [inline src/transform/src/predicate_pushdown.rs:937]: I believe the following implementation of this method is a bit more readable and robust (no recursive methods).
Also, if you slightly generalize the signature you can pass and handle all filter `predicates: &Vec<MirScalarExpr>` in one call and build the `memo` table only once for all `predicates`.

```rust
fn inline_if_not_too_big(
    expr: &MirScalarExpr,
    input_arity: usize,
    map_exprs: &Vec<MirScalarExpr>,
) -> Result<Option<MirScalarExpr>, RecursionLimitError> {
    let size_limit = 1000;
    let mut memo = BTreeMap::<usize, (MirScalarExpr, usize)>::new();

    // Recursively determine the support of `expr` produced by `map_exprs`
    // that needs to be memoized.
    let memoize_columns = {
        let mut support = BTreeSet::new();

        // Seed with `map_exprs` support in `expr`.
        expr.visit_post(&mut |e| {
            if let MirScalarExpr::Column(c) = e {
                if *c >= input_arity {
                    support.insert(*c);
                }
            }
        })?;

        // Compute transitive closure with `map_exprs` in support.
        let mut workset = support.iter().cloned().collect::<Vec<_>>();
        let mut buffer = vec![];
        while !workset.is_empty() {
            // Swap the (empty) `drained` buffer with the `workset`.
            std::mem::swap(&mut workset, &mut buffer);
            // Drain the `buffer` and update `support` and `workset`.
            for c in buffer.drain(..) {
                map_exprs[c - input_arity].visit_post(&mut |e| {
                    if let MirScalarExpr::Column(c) = e {
                        if *c >= input_arity {
                            if !support.insert(*c) {
                                workset.push(*c);
                            }
                        }
                    }
                })?;
            }
        }
        support
    };

    // Populate the memo table in ascending column order (which respects the
    // dependency order of map_exprs references). Break early if memoization
    // fails for one of the columns in `memoize_columns`.
    for c in memoize_columns.iter() {
        let mut new_expr = map_exprs[*c - input_arity].clone();
        let mut new_size = 0;
        new_expr.visit_mut_post(&mut |expr| {
            new_size += 1;
            if let MirScalarExpr::Column(c) = expr {
                if *c >= input_arity && new_size <= size_limit {
                    let (m_expr, m_size) = &memo[c]; // Asserted by the break below.
                    *expr = m_expr.clone();
                    new_size += m_size - 1; // Adjust for the +1 above.
                }
            }
        })?;

        if new_size <= size_limit {
            memo.insert(*c, (new_expr, new_size));
        } else {
            break;
        }
    }

    // Try to resolve expr against the memo table.
    if memo.len() < memoize_columns.len() {
        Ok(None) // We couldn't memoize all map expressions within the given limit.
    } else {
        let mut new_expr = expr.clone();
        let mut new_size = 0;
        new_expr.visit_mut_post(&mut |expr| {
            new_size += 1;
            if let MirScalarExpr::Column(c) = expr {
                if *c >= input_arity && new_size <= size_limit {
                    let (m_expr, m_size) = &memo[c];
                    *expr = m_expr.clone(); // Asserted by the enclosing if condition.
                    new_size += m_size - 1; // Adjust for the +1 above.
                }
            }
        })?;

        if new_size <= size_limit {
            Ok(Some(new_expr)) // We managed to stay within the limit.
        } else {
            Ok(None) // Limit exceeded.
        }
    }
}
```

**Review by aalexandrov — CHANGES_REQUESTED** [review]
Looks good, but the `inline_if_not_too_big` is a bit hard to follow. I left a comment with an alternative implementation which passes the changed `*.slt` files locally and seems a bit more direct.

File: src/transform/src/predicate_pushdown.rs (hunk 1 above)
Comment (ggevay) [reply src/transform/src/predicate_pushdown.rs:937]: Thank you! I adopted this implementation. Even though it's a bit longer, it might indeed be more readable.

Comment (ggevay) [reply src/transform/src/predicate_pushdown.rs:937]: I didn't do the bundling together of several predicates, because then it becomes complicated to handle the case when one predicate would run over the limit, but there are some other smaller predicates where the inlining should succeed.

Comment (ggevay) [reply src/transform/src/predicate_pushdown.rs:945]: I added a code comment about this optimization:
```
    /// OOO: (Optimizer Optimization Opportunity) This function might do work proportional to the
    /// total size of the Map expressions. We call this function for each predicate above the Map,
    /// which will be kind of quadratic, i.e., if there are many predicates and a big Map, then this
    /// will be slow. We could instead pass a vector of Map expressions and call this fn only once.
    /// The only downside would be that then the inlining limit being hit in the middle part of this
    /// function would prevent us from inlining any predicates, even ones that wouldn't hit the
    /// inlining limit if considered on their own.
```
I'd propose to use `OOO` to mark also other optimization opportunities in the optimizer, so that when we need to optimize the optimizer some time in the next few years, we can get back to these.

**Review by aalexandrov — APPROVED** [review]
